ballerina/websocket_service.bal

```diff
@@ -54,67 +54,80 @@ isolated service class WsService {
+                        closeConnection(caller, 4429, "Too many initialisation requests");
                         return;
                     }
-                    if self.activeConnections.indexOf(connectionId) !is () {
-                        closeConnection(caller, 4409, string `Subscriber for ${connectionId} already exists`);
-                        return;
+                    check caller->writeMessage({"type": WS_ACK});
+                    self.initiatedConnection = true;
+                }
+            }
+            WS_SUBSCRIBE|WS_START => {
+                lock {
+                    if self.customHeaders.hasKey(WS_SUB_PROTOCOL) {
+                        if !self.initiatedConnection {
+                            closeConnection(caller, 4401, "Unauthorized");
+                            return;
+                        }
+                        if self.activeConnections.indexOf(connectionId) !is () {
+                            closeConnection(caller, 4409, string `Subscriber for ${connectionId} already exists`);
+                            return;
+                        }
+                        self.activeConnections.push(connectionId);
                     }
-                    self.activeConnections.push(connectionId);
                 }
+                return self.handleSubscriptionRequest(caller, connectionId, wsPayload);
             }
-            parser:OperationNode|json node = validateSubscriptionPayload(wsPayload, self.engine);
-            if node is parser:OperationNode {
-                check executeOperation(self.engine, self.context, self.schema, self.customHeaders, caller,
-                                       connectionId, node);
-            } else {
-                check sendWebSocketResponse(caller, self.customHeaders, WS_ERROR, node, connectionId);
+            WS_STOP|WS_COMPLETE => {
+                lock {
+                    _ = self.activeConnections.remove(<int>self.activeConnections.indexOf(connectionId));
+                    self.initiatedConnection = false;
+                }
+                check sendWebSocketResponse(caller, self.customHeaders, WS_COMPLETE, null, connectionId);
                 closeConnection(caller);
             }
-        } else if wsType == WS_STOP || wsType == WS_COMPLETE {
-            lock {
-                _ = self.activeConnections.remove(<int>self.activeConnections.indexOf(connectionId));
-                self.initiatedConnection = false;
-            }
-            check sendWebSocketResponse(caller, self.customHeaders, WS_COMPLETE, null, connectionId);
-            closeConnection(caller);
-        } else if wsType == WS_PING {
+            WS_PING => {
             check caller->writeMessage({"type": WS_PONG});
-        } else if wsType == WS_PONG {
-            check caller->writeMessage({"type": WS_PING});
+            }
+            WS_PONG => {
+                check caller->writeMessage({"type": WS_PING});
```

Comment:
we don't have to write a ping message here

Reply:
Did not do that change since you are working on the subprotocol fix.